src/frontend/src/flows/addDevice/welcomeView/registerTentativeDevice.ts

```diff
@@ -1,174 +1,168 @@
+  );
+};
 
-  aliasInput.onkeypress = (e) => {
-    // submit if user hits enter
-    if (e.key === "Enter") {
-      e.preventDefault();
-      continueButton.click();
+/** Create new WebAuthn credentials */
+const createDevice = async ({
+  userNumber,
+  connection,
+}: {
+  userNumber: bigint;
+  connection: Connection;
+}): Promise<WebAuthnIdentity | Error> => {
+  const existingAuthenticators = await connection.lookupAuthenticators(
+    userNumber
+  );
+  let newDevice: WebAuthnIdentity;
+  try {
+    newDevice = await WebAuthnIdentity.create({
+      publicKey: creationOptions(existingAuthenticators),
+    });
+  } catch (error: unknown) {
+    if (error instanceof Error) {
+      return error;
+    } else {
+      return new Error(unknownToString(error, "unknown error"));
     }
-  };
+  }
+  return newDevice;
+};
 
-  continueButton.onclick = async () => {
-    if (aliasInput.value === "") {
-      toggleErrorMessage("tentativeDeviceAlias", "invalidAliasMessage", true);
-      return;
-    }
+type AddDeviceSuccess = Extract<
+  AddTentativeDeviceResponse,
+  { added_tentatively: unknown }
+>;
 
-    let newDevice: WebAuthnIdentity;
-    try {
-      newDevice = await WebAuthnIdentity.create({
-        publicKey: creationOptions(existingAuthenticators),
-      });
-    } catch (error: unknown) {
+/** Add the device tentatively to the canister */
+export const addTentativeDevice = async ({
+  userNumber,
+  connection,
+  device,
+}: {
+  userNumber: bigint;
+  connection: Connection;
+  device: Omit<DeviceData, "origin">;
+}): Promise<AddDeviceSuccess> => {
+  // Try to add the device tentatively, retrying if necessary
+  for (;;) {
+    const result = await withLoader(() =>
+      connection.addTentativeDevice(userNumber, device)
+    );
+
+    if ("another_device_tentatively_added" in result) {
+      // User already added a device so we show an error and abort
       await displayError({
-        title: "Error adding new device",
-        message: "Unable to register new WebAuthn Device.",
-        detail: error instanceof Error ? error.message : JSON.stringify(error),
+        title: "Tentative Device Already Exists",
+        message:
+          'The "add device" process was already started for another device. If you want to add this device instead, log in using an existing device and restart the "add device" process.',
         primaryButton: "Ok",
       });
       // TODO L2-309: do this without reload
-      window.location.reload();
-      return;
+      return window.location.reload() as never;
+    }
+
+    if ("device_registration_mode_off" in result) {
+      // User hasn't started the "add device" flow, so we offer to enable it and retry, or cancel
+      const res = await deviceRegistrationDisabledInfo(userNumber);
+      if (res === "canceled") {
+        // TODO L2-309: do this without reload
+        return window.location.reload() as never;
+      }
+
+      if (res === "retry") {
+        continue;
```

Comment:
it's copied or there is a way to avoid the use of`continue` keywords?

for sure it works but from experience, often had bugs in piece of codes that use such keywords, therefore I like to avoid `continue`. 

totally opiniated therefore totally forgetable comment, just speaking out at loud.

Reply:
no you're completely right; just wasn't sure how to expressed this any other way. The behavior is the same as before, except the logic was spread across 2 modules and many different functions which obfuscated the issue 😬